lib/core/macros.nim

```diff
@@ -1344,6 +1348,76 @@ proc copy*(node: NimNode): NimNode {.compileTime.} =
   ## An alias for copyNimTree().
   return node.copyNimTree()
 
+type GenAstOpt* = enum
+  kDirtyTemplate,
+    # When set, uses a dirty template in implementation of `genAst`. This
+    # is occasionally useful as workaround for issues such as #8220, see
+    # `strformat limitations <strformat.html#limitations>`_ for details.
+    # Default is unset, to avoid hijacking of uncaptured local symbols by
+    # symbols in caller scope.
+  kNoNewLit,
+    # don't call call newLit automatically in `genAst` capture parameters
+
+macro genAstOpt*(options: static set[GenAstOpt], args: varargs[untyped]): untyped =
+  ## Accepts a list of captured variables `a=b` or `a` and a block and returns the
+  ## AST that represents it. Local `{.inject.}` symbols (e.g. procs) are captured
+  ## unless `kDirtyTemplate in options`.
+  runnableExamples:
+    macro fun(a: bool, b: static bool): untyped =
+      let c = false # doesn't override parameter `c`
+      var d = 11 # var => gensym'd
+      proc localFun(): auto = 12 # proc => inject'd
+      genAst(a, b, c = true):
+        # echo d # not captured => gives `var not init`
+        (a, b, c, localFun())
+    doAssert fun(true, false) == (true, false, true, 12)
+
+  let params = newTree(nnkFormalParams, newEmptyNode())
+  let pragmas =
+    if kDirtyTemplate in options:
+      nnkPragma.newTree(ident"dirty")
+    else:
+      newEmptyNode()
+
+  template newLitMaybe(a): untyped =
+    when type(a) is NimNode: a
```

Comment:
It doesn't print a warning yet, but it will. ``typeof`` is the correct keyword to use here.

Reply:
done